tensorboard/plugins/projector/vz_projector/data.ts

```diff
@@ -370,6 +378,84 @@ export class DataSet {
     });
   }
 
+  /** Runs UMAP on the data. */
+  projectUmap(nComponents: number, nNeighbors: number, stepCallback: (iter: number) => void) {
```

Comment:
So, after reading through this logic, the control flow still isn’t clear
to me. Why is `requestAnimationFrame` only called once, outside of
`step`, while `step` itself recurs synchronously? The `nEpochs`
variable, which appears to hold the target (maximum) number of epochs,
is initialized to `1`, but later mutated from within a _different_
asynchronous callback to the result of `initializeFit`, so which value
is used when?

Is there some way that we can restructure this to make the control flow
obvious from the structure of the code? For instance, if the initial
value of `1` for `nEpochs` is supposed to be an unobservable state, then
could we write something like

```typescript
async projectUmap(/* … */) {
  const nearest = await /* compute knnComputation */;
  const nEpochs = await util.runAsyncTask('Initializing UMAP…', () => {
    // …
    return this.umap.initializeFit(/* … */);
  });
  return new Promise((resolve) => {
    const step = () => {
      // …
      requestAnimationFrame(step);
      // …
      if (done) resolve();
    };
    requestAnimationFrame(step);
  });
}
```

(I think that I’ve reached my short-term saturation on understanding
this function and allotting time to it, so I’ll come back to see what
you come up with. :-) )

Reply:
Great, thanks for such a thorough review.

Yeah, I wasn't particularly pleased with how I wound up doing this, it was pretty much hacked together to work like the t-SNE implementation before I had a really good grasp on how the application worked together. Anyway, I've refactored it completely to make it a bit more clear what's going on and why it's going on that way.

Reply:
Thank you! This is much easier for me to read.